src/System.Diagnostics.Tracing.Telemetry/src/System/Diagnostics/Tracing/TelemetryListener.cs

```diff
@@ -0,0 +1,249 @@
+        /// as well as those that occured in the past, and all future Listeners created in the future. 
+        /// </summary>
+        public static event Action<TelemetryListener> AllListeners
+        {
+            add
+            {
+                lock (DefaultListener)
+                {
+                    s_allListenersCallback = (Action<TelemetryListener>)Delegate.Combine(s_allListenersCallback, value);
+
+                    // Call back for each existing listener on the new callback.  
+                    for (TelemetryListener cur = s_allListeners; cur != null; cur = cur.m_next)
+                        value(cur);
+                }
+            }
+            remove
+            {
+                s_allListenersCallback = (Action<TelemetryListener>)Delegate.Remove(s_allListenersCallback, value);
+            }
+        }
+
+        // Subscription implementation 
+        /// <summary>
+        /// Add a subscriber (Observer).  If 'IsEnabled' == null (or not present), then the Source's IsEnabled 
+        /// will always return true.  
+        /// </summary>
+        virtual public IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Predicate<string> isEnabled)
+        {
+            // If we have been disposed, we silently ingore any subscriptions.  
+            if (m_disposed)
+                return new Subscription() { Owner = this };
+
+            Subscription newSubscription = new Subscription() { Observer = observer, IsEnabled = isEnabled, Owner = this, Next = m_subscriptions };
+            while (Interlocked.CompareExchange(ref m_subscriptions, newSubscription, newSubscription.Next) != newSubscription.Next)
+                newSubscription.Next = m_subscriptions;
+            return newSubscription;
+        }
+        public IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer)
+        {
+            return Subscribe(observer, null);
+        }
+
+        /// <summary>
+        /// Make a new TelemetryListener, it is a NotificationSource, which means the returned result can be used to 
+        /// log notifications, but it also has a Subscribe method so notifications can be forwarded
+        /// arbitrarily.  Thus its job is to forward things from the producer to all the listeners
+        /// (multi-casting).    Generally you should not be making your own TelemetryListener but use the
+        /// TelemetryListener.Default, so that notifications are as 'public' as possible.  
+        /// </summary>
+        public TelemetryListener(string name)
+        {
+            Name = name;
+            // To avoid allocating an explicit lock object I lock the Default TelemetryListener.   However there is a 
+            // chicken and egg problem because I need to call this code to initialize Default.   
+            var lockObj = DefaultListener;
+            if (lockObj == null)
+                lockObj = this;
```

Comment:
Ah, I see.  The only time this should happen is when creating the default listener, so this will become default listener, and we need to synchronize because of the dance done with s_allListeners in the lock block.  I suggest adding a Debug.Assert to ensure this is the case, e.g.
```C#
if (lockObj == null)
{
    Debug.Assert(name == "TelemetryListener.Default");
    lockObj = this;
}
```